### ISSUE TITLE
Default the Animation Import Framerate to Match the Source Framerate

### ISSUE THREAD

**Issue by jitspoe** [godotengine/godot-proposals#430]
**Describe the project you are working on:**
2.5D Platformer (Fist of the Forgotten) using 3D assets and animations.

**Describe the problem or limitation you are having in your project:**
When importing models, I have to make sure I specify the same fps for importing animations as I have in my blender project, otherwise my animations lose frames or have distorted lerps.

**Describe how this feature / enhancement will help you overcome this problem or limitation:**
Defaulting the import framerate to use whatever is in the exported file will ensure the keyframes match 1:1 with the original animation and not result in distorted/mushy animation if the import settings don't match exactly (aka: it should just work without having to mess with potentially unknown settings).  I had a number of animation issues when I first started using Godot, and my 60fps animations being converted to 15fps and losing frames was one of them.  If Godot just matched the 60fps framerate by default, it would have saved a lot of time trying to debug why my animations were distorted.

**Show a mock up screenshots/video or a flow diagram explaining how your proposal will work:**
![image](https://user-images.githubusercontent.com/7751182/73607132-5f95ba80-4580-11ea-99a7-50a518937f52.png)

**Describe implementation detail for your proposal (in code), if possible:**
I noticed the collada importer had this code:

```
		if (nm.anim_tracks.size() == 1) {
			//use snapshot keys from anim track instead, because this was most likely exported baked
			const Collada::AnimationTrack &at = collada.state.animation_tracks[nm.anim_tracks.front()->get()];
			snapshots.clear();
			for (int i = 0; i < at.keys.size(); i++)
				snapshots.push_back(at.keys[i].time);
		}
```

I'm suggesting this be the default behavior.  Just change the anim_tracks.size() check to `if (!override_fps) {`

Also, there are some slight floating point precision epsilons that would need to be added for the frame timing being slightly off.  In the collada importer `Transform interp = c < 0.001 ? src : src.interpolate_with(dst, c);` doesn't account for c being slightly off the other way (0.999...).  Not sure about the other importers.

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
I don't think so, since it's a core engine feature.

**Is there a reason why this should be core and not an add-on in the asset library?:**
Importing is a core feature.  I think this is something that should "just work" without having to tinker with input settings.

**Comment by fire** [comment]
From trials with 3d models, the default should at least be 30 fps.

**Comment by Calinou** [comment]
Is there even an use in overriding the import FPS?

**Comment by fire** [comment]
Well, 15 fps breaks a large portion of my animations. 30 fps is standard. 60 fps is good, but requires double the keyframes of 30 if linearly baked.

**Comment by Calinou** [comment]
@fire What I mean is that I'd expect the imported model to come with a suitable animation FPS out of the bopx.

**Comment by golddotasksquestions** [comment]
@Calinou I would think the override exists in case you have animations from different sources with coming in with different framerates. So you can still open and use them in Godot without having to go into your 3D animation software to check and correct framerates, no?

**Comment by jitspoe** [comment]
I'm honestly not sure why you'd want to override the framerate of the animation when importing it.  I'm merely suggesting keeping it because that's how it currently works and somewhere out there, somebody might be using it.  Not sure why it was designed like this in the first place, though.  You'd think it would just import the animation frames as they were in the source file.

With optimization enabled, even if you have a high framerate animation, it should eliminate keyframes that follow a near-linear path, right?

**Comment by fire** [comment]
Due to floating point precision equality problems, there needs to be code that fits a beizer curve through the linear points. This is why it won't eliminate keyframes in practice.

**Comment by jitspoe** [comment]
So I took a stab at this because I got frustrated by something importing at the wrong framerate again: https://github.com/jitspoe/godot/commit/c549f69b496406a8200cdcb0a9bb5277e7ba0566

Not really familiar with shape keys, so I didn't touch that yet.

**Comment by fire** [comment]
Blend shapes are animations that go from 0 to 1. It would be affected by frame rate the same way.

**Comment by fire** [comment]
I was able too get consensus to boost the default framerate to 30fps.

**Comment by prominentdetail** [comment]
When I import a gltf, the frames get repositioned.. 
I have a frame at like 0.0001 time difference from another frame and it ends up getting repositioned to have a much larger time between frames when imported. This breaks my animation.

How can I get the frames to retain their correct timing?

**Comment by KaletheQuick** [comment]
This problem still persists. I double checked the various settings and didn't find anywhere to set a custom default import framerate. Is there any path to solving this issue?

**Comment by Calinou** [comment]
> and didn't find anywhere to set a custom default import framerate

You can change the default value of any import option in the **Import Defaults** tab of the Project Settings.

**Comment by KaletheQuick** [comment]
>  **Import Defaults** tab of the Project Settings.
Thanks! I was looking for GLTF or something, found it under 'scenes' 
:D

**Comment by fire** [comment]
tl;dr I'm not sure how to do this, we're basically taking the keyframes and baking them at intervals.

**Comment by prominentdetail** [comment]
Why can't it just import only the keyframes that are in the animation, instead of generating extra keyframes?

**Comment by fire** [comment]
1. gltf cubic isn't the same as godot engine cubic.
2. Godot Engine animations stores frames by times

**Comment by jitspoe** [comment]
I'm not sure how the gltf format stores keyframes, but would it not be possible to grab all the exact keyframes as [time, value] pairs then for the rest use the interpolated values at the rate specified in the import settings?  That would allow for importing anims that have keyframes like .0001s apart without having to import at some obscene framerate.

**Comment by LordMcMutton** [comment]
Does the gltf format store the framerate? If so, could Godot throw a warning that there's a mismatch between the import framerate and the exported framerate?

**Comment by jitspoe** [comment]
Could we bake at a set interval to cover the interpolation but also force keyframes where they exist in GLTF?  For example, bake at a set 30fps or whatever the import setting is set to, but if the GLTF file is at 100fps with some explicit keyframes that differ from the specified rate, store those, too.
